design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,146 @@
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- As a VM/VMI administrator, I would like to create a VM that is able
+  to increase its network interfaces count in the future.
+  I would explicitly define the amount of potential network interfaces
+  by providing a value through the VM/VMI Spec.
+
+### API Extensions
+The VMI Spec will be extended to include a new domain resource name: `kubevirt.io/interface`.
+It will be available as a domain-resources-requests entry.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 0 and 32.
+- Should be accepted only as `requests` (not as `limits`).
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+> **Note**: These are API level validations. The backend (e.g. hypervisor)
+> may have other validations/limitation.
+
+#### VMI Spec example
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-request-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+    resources:
+      requests:
+        kubevirt.io/interface: 5
+    rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the resources-requests-kubevirt.io/interface and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
+
+#### Network Interface Placeholder
+The interface placeholder is a temporary dummy definition of an interface in the domain spec
+of libvirt. It is used to trigger libvirt to auto-generate dependent devices (e.g. PCIe controllers).
+
+After the domain configuration is constructed, these placeholders are removed.
+
+This is an example of such a placeholder:
+```xml
+<interface type='ethernet'>
+  <target dev='placeholder-1' managed='no'/>
+  <model type='virtio-non-transitional'/>
+</interface>
+```
+
+## Appendix A: Alternatives
```

Comment:
I hope that @andreabolognani can help out with phrasing these with more content.

Reply:
In any case, I suggest to do this as a follow up.

Reply:
Taking a stab at this.

---

- While using the same suggested API towards the user, construct the required controllers directly (and not through network interface placeholders).

  Would require introducing knowledge about machine type-specific details into KubeVirt: for example, `q35` requires `pcie-root-port` controllers to be added but `pc` doesn't, and would in fact break if those were present.

- Always define the maximum number of resources by default (or as a boolean input).

  Each additional PCI controller introduces some overhead in terms of memory usage and guest OS boot time, so minimizing their amount is beneficial. That said, the exact overhead hasn't been measured, and it could turn out not to be a deal-breaker compared to the baseline resource usage.

- Use a PCI bridge.

  Would require implementing PCI address allocation logic in KubeVirt, as libvirt will normally refuse to use the conventional PCI slots provided by a `pci-bridge` controller for devices on a PCI Express-based machine type such as `q35`. The logic would have to be used for all devices, not just network interfaces, in order to avoid clashes with the addresses allocated by libvirt.

---

Feel free to use it either as-is or as a starting point.

Reply:
I updated the proposal with this new info.

Reply:
Thank you for that.

> But it would be unfortunate to load the domain with PCI controllers which have impact on boot time and memory when 98% if the time it is not needed.

The boolean defeats this argument IMO; you don't need / want hotplug, don't set the boolean.

But I see the point the same approach can be achieved with the resource number you're proposing, and I think this might be more re-usable across the board for similar scenarios (CPU / memory hotplug).